### PR TITLE
Added test cases for dassert_file_extension()

### PR DESCRIPTION
**Pull request by dreamcatcher0101**
### This PR resolved the issue [#466](https://github.com/sorrentum/sorrentum/issues/466) 

**How it was resolved?**
- Added case to check if filename has given extension.
> Modified dassert_file_extension() to get correct extension
- Added case to check if filename has one of given extensions.
- Added case to check if assertion is raised when filename doesn't have one of given extensions.

### CODE REVIEW

**Comment by dreamcatcher0101** [comment]
Please review my code.

File: helpers/test/test_dbg.py [hunk 1]
```diff
@@ -624,3 +624,42 @@ def test3(self) -> None:
...
+
+# #############################################################################
+
+
+class Test_dassert_file_extension1 (hunitest.TestCase):
+    def test1(self) -> None:
+        """
+        Check if file has given extension.
+        """
+        extension = r".csv.gz"
```
Comment (PomazkinG) [inline helpers/test/test_dbg.py:637]: why do we need `r` everywhere? can we remove?

File: helpers/hdbg.py [hunk 2]
```diff
@@ -806,7 +806,7 @@ def dassert_file_extension(
     # Make sure extension starts with .
     extensions = [f".{e}" if not e.startswith(".") else e for e in extensions]
     # Check.
-    act_ext = os.path.splitext(file_name)[-1].lower()
+    act_ext = file_name[file_name.find("."):]
```
Comment (PomazkinG) [inline helpers/hdbg.py:809]: an example from the real-life: typically we use also use `.` to separate pieces of information in the file name. 

e.g., `order.20231701_120057.20231701_13000.csv.gz`. here we separate timestamps with `.`. According to your code the resulting extension will be `20231701_120057.20231701_13000.csv.gz`

please modify the algo + add a separate test for `order.20231701_120057.20231701_13000.csv.gz`

**Comment by dreamcatcher0101** [comment]
Hi. @PomazkinG 
Please review again.

File: helpers/hdbg.py [hunk 3]
```diff
@@ -308,9 +308,13 @@ def dassert_in(
     *args: Any,
     only_warning: bool = False,
 ) -> None:
-    cond = value in valid_values
+    cond = False
+    fileExtension = act_ext = os.path.splitext(str(value))[-1].lower()
```
Comment (PomazkinG) [inline helpers/hdbg.py:312]: not sure it's a good idea to modify `dassert_in()` just for checking file extensions because it's a general function (grep for it)

1. It will lead to a global regression break
2. the function is not designed for file extensions exclusively 

what I would do:
1. Modify only `dassert_file_extension()`
2. Maybe keep the existing logic
3. Add a special case: if a file name ends with `.gz` the extension has the pattern `.{extension}.gz` otherwise the pattern is `.{extension}`
   - Probably `.gz` is the only popular extension that has 2 `.`, e.g., `.csv.gz` so treating it as a special case is fine IMO
   - In simple words the idea is: extension is the set of chars after a dot, but if a file name ends with `.gz` the extension is the set of chars starting from the dot that goes before `.gz`

Comment (PomazkinG) [reply helpers/hdbg.py:312]: E.g., `.csv` -> extension is `.csv`, `.csv.gz` -> extension is `.csv.gz`

**Comment by DanilYachmenev** [comment]
@dreamcatcher0101 pls change PR labels accordingly when you want your PR to be reviewed
https://github.com/sorrentum/sorrentum/blob/master/docs/First_review_process.md#ask-for-reviews

File: helpers/hdbg.py [hunk 4]
```diff
@@ -807,6 +807,9 @@ def dassert_file_extension(
     extensions = [f".{e}" if not e.startswith(".") else e for e in extensions]
     # Check.
     act_ext = os.path.splitext(file_name)[-1].lower()
+    if act_ext == ".gz":
+        act_ext = os.path.splitext(file_name[:-3])[-1].lower() + ".gz"
```
Comment (DanilYachmenev) [inline helpers/hdbg.py:811]: Let's try to make logic a bit clearer with smth like
```
name, act_ext = os.path.splitext(file_name.lower())
if act_ext == ".gz"
    # Get a file extension that is preceding gzip extension. 
    ext = os.path.splitext(name)[-1]
    act_ext = ext + act_ext
```

File: helpers/test/test_dbg.py [hunk 5]
```diff
@@ -624,3 +624,51 @@ def test3(self) -> None:
...
+        """
+        Check if file has one of the given extensions.
+        """
+        extension = [".csv.gz", ".csv"]
+        filename = "extensionFile.csv"
+        hdbg.dassert_file_extension(filename, extension)
+    
+    def test3(self) -> None:
+        """
+        Case of filename that have incorrect extension with given extensions.
```
Comment (DanilYachmenev) [inline helpers/test/test_dbg.py:651]: -> `Check that assertion is raised if filename has none of the given extensions.`

File: helpers/test/test_dbg.py [hunk 6]
```diff
@@ -624,3 +624,51 @@ def test3(self) -> None:
...
+'.txt' in '['.csv.gz', '.csv']'
+Invalid extension '.txt' for file 'txtFile.txt'
+################################################################################
+        """
+        self.assert_equal(act, exp)
+
+    def test4(self) -> None:
+        """
+        Edge Case of multi dots.
+        e.g. filename = "order.20231701_120057.20231701_13000.csv.gz"
```
Comment (DanilYachmenev) [inline helpers/test/test_dbg.py:670]: -> `Check that a file with multiple dots in its name passes the assertion.`

[the remark this answers is not in the document]
Comment (dreamcatcher0101) [reply helpers/test/test_dbg.py:670]: Hi, @DanilYachmenev .
@PomazkinG asked me to add a separate test for "order.20231701_120057.20231701_13000.csv.gz", so I added it.
Is it ok to remove?

[the remark this answers is not in the document]
Comment (DanilYachmenev) [reply helpers/test/test_dbg.py]: Oh, I menat remove the 2nd docstring line `e.g. filename = "order.20231701_120057.20231701_13000.csv.gz"`
keep the test

File: helpers/hdbg.py [hunk 7]
```diff
@@ -807,6 +807,11 @@ def dassert_file_extension(
     extensions = [f".{e}" if not e.startswith(".") else e for e in extensions]
     # Check.
     act_ext = os.path.splitext(file_name)[-1].lower()
+    if act_ext == ".gz":
+        # Get a file extension that is preceding gzip extension. 
+        # get the set of chars starting from the dot that goes before '.gz'
+        ext = os.path.splitext(file_name[:-3])[-1].lower()
+        act_ext = ext + act_ext
```
Comment (DanilYachmenev) [inline helpers/hdbg.py:814]: see my suggestion for these lines
IMO `os.path.splitext(file_name[:-3])[-1].lower()` is quite vague, takes more time to comprehend

also you use `.lower()` twice here which is not DRY

File: helpers/hdbg.py (hunk 4 above)
Comment (DanilYachmenev) [reply helpers/hdbg.py]: https://github.com/sorrentum/sorrentum/pull/477#discussion_r1279106232

**Comment by DanilYachmenev** [comment]
Press `Resolve conversation` after you implement the comment

File: helpers/hdbg.py (hunk 4 above)
Comment (DanilYachmenev) [reply helpers/hdbg.py]: Also pls run the tests after you implement all the changes and confirm that they pass

File: helpers/test/test_dbg.py [hunk 8]
```diff
@@ -624,3 +624,51 @@ def test3(self) -> None:
...
+        filename = "txtFile.txt"
+        with self.assertRaises(Exception) as cm:
+            hdbg.dassert_file_extension(filename, extension)
+        act = str(cm.exception)
+        exp = """
+################################################################################
+* Failed assertion *
+'.txt' in '['.csv.gz', '.csv']'
+Invalid extension '.txt' for file 'txtFile.txt'
+################################################################################
```
[the remark this answers is not in the document]
Comment (dreamcatcher0101) [reply helpers/test/test_dbg.py]: @DanilYachmenev 
When I try to indent it on the same level where quotes start, the tab are reflected to the output and the result fails.
Is it ok to implement it as one line? like this...
exp = """\n################################################################################\n* Failed assertion *\n'.txt' in '['.csv.gz', '.csv']'\nInvalid extension '.txt' for file 'txtFile.txt'\n################################################################################\n"""
I think it's a little bit weird, but it's my solution.

[the remark this answers is not in the document]
Comment (DanilYachmenev) [reply helpers/test/test_dbg.py]: add `r` before right the quotes, like
```
exp = r"""
...
this should solve the problem
```

[the remark this answers is not in the document]
Comment (dreamcatcher0101) [reply helpers/test/test_dbg.py]: @DanilYachmenev 
I couldn't solve the problem with adding "r".
The tabs are still reflected to the output.

[the remark this answers is not in the document]
Comment (DanilYachmenev) [reply helpers/test/test_dbg.py]: you also need to fix the last line ->
```
self.assert_equal(act, exp, purify_text=True, fuzzy_match=True)
``` 

and use `r"""` for the output ofc

File: helpers/test/test_dbg.py [hunk 9]
```diff
@@ -624,3 +625,51 @@ def test3(self) -> None:
         message 'hello world'
         """
         self.assert_equal(act, exp, purify_text=True, fuzzy_match=True)
+
+
+# #############################################################################
+
+
+class Test_dassert_file_extension1 (hunitest.TestCase):
```
Comment (PomazkinG) [inline helpers/test/test_dbg.py:633]: nit: remove space before `(...)`

File: helpers/test/test_dbg.py [hunk 10]
```diff
@@ -624,3 +625,51 @@ def test3(self) -> None:
...
+            hdbg.dassert_file_extension(filename, extensions)
+        act = str(cm.exception)
+        exp = r"""
+        ################################################################################        
+        * Failed assertion *
+        '.txt' in '['.csv.gz', '.csv']'
+        Invalid extension '.txt' for file 'txtFile.txt'
+        ################################################################################
+        """
+        self.assert_equal(act, exp, purify_text=True, fuzzy_match=True)
```
Comment (PomazkinG) [inline helpers/test/test_dbg.py:667]: nit: maybe not use the `purify` switch here?

in this case we can:
1) remove the `set_user_name()`
2) the `hsystem` import

it does not seem to do its job here

File: devops/docker_build/create_users.sh [hunk 11]
```diff
@@ -12,8 +12,8 @@ useradd -u 501 $OPTS user_501
 usermod -aG docker user_501
 
 # Linux users.
-# We start from id=1001, the upper bound can be changed if needed.
-for current_linux_id in {1001..1050}
+# We start from id=1000 because IDs below 1000 are usually reserved for system users and groups, the upper bound can be changed if needed.
```
Comment (PomazkinG) [inline devops/docker_build/create_users.sh:15]: no need to it in this PR, the fix belongs to https://github.com/sorrentum/sorrentum/pull/465, please revert

File: helpers/hdbg.py [hunk 12]
```diff
@@ -806,7 +806,11 @@ def dassert_file_extension(
     # Make sure extension starts with .
     extensions = [f".{e}" if not e.startswith(".") else e for e in extensions]
     # Check.
-    act_ext = os.path.splitext(file_name)[-1].lower()
+    name, act_ext = os.path.splitext(file_name)
```
Comment (PomazkinG) [inline helpers/hdbg.py:809]: nit: remove the TODO at line 793, since your PR addresses it

File: helpers/hdbg.py [hunk 13]
```diff
@@ -806,7 +806,11 @@ def dassert_file_extension(
...
+    if act_ext == ".gz":
+        # Get a file extension that is preceding gzip extension. 
+        ext = os.path.splitext(name)[-1]
+        act_ext = (ext + act_ext).lower()
```
Comment (PomazkinG) [inline helpers/hdbg.py:813]: nit:
```
# Concatenate with the preceding extension, e.g., `.csv.gz`.
```

**Review by PomazkinG — COMMENTED** [review]
Nice, only nits and we merge

**Review by DanilYachmenev — COMMENTED** [review]
lgtm

just pls confirm that the tests pass (especially `test3` since we had problems with it)

**Comment by dreamcatcher0101** [comment]
> lgtm
> 
> just pls confirm that the tests pass (especially `test3` since we had problems with it)

All the tests are passed.

**Review by PomazkinG — APPROVED** [review]
LGTM
